### PR TITLE
feat: add functions to extract base coordinates of Feature

### PR DESCRIPTION
**Pull request by mnsmar**
Functions for feat package to extract the base coordinates for a feat.Feature.

### CODE REVIEW

File: feat/feature_test.go [hunk 1]
```diff
@@ -288,3 +288,67 @@ func (s *S) TestOrientationWithin(c *check.C) {
...
+	for _, t := range baseCoordsTests {
+		c.Check(feat.BaseStartOf(t.f), check.Equals, t.baseStart)
+		c.Check(feat.BaseEndOf(t.f), check.Equals, t.baseEnd)
+		c.Check(feat.BasePosOf(t.f, t.pos), check.Equals, t.basePos)
+	}
+
+	// Check we detect cycles.
+	var cycle ori
+	cycle.loc = &cycle
+	c.Check(func() { feat.BasePosOf(cycle, 10) },
```
Comment (kortschak) [inline feat/feature_test.go]: One line.

File: feat/feature.go [hunk 2]
```diff
@@ -105,6 +105,38 @@ type Collection interface {
...
+// location, which may be the feature itself if it has a nil Location.
+// BaseEndOf will panic if the feature chain is deeper than 1000 links.
+func BaseEndOf(f Feature) int {
+	return BasePosOf(f.Location(), f.End())
+}
+
+// BasePosOf converts a position that is defined on a given feature to a
+// position relative to the deepest location of this feature and returns it.
+// BasePosOf will panic if the feature chain is deeper than 1000 links.
+func BasePosOf(f Feature, pos int) int {
```
Comment (kortschak) [inline feat/feature.go]: s/pos/position/

File: feat/feature.go [hunk 3]
```diff
@@ -105,6 +105,38 @@ type Collection interface {
...
+}
+
+// BaseEndOf returns the end of the given feature relative to the deepest
+// location, which may be the feature itself if it has a nil Location.
+// BaseEndOf will panic if the feature chain is deeper than 1000 links.
+func BaseEndOf(f Feature) int {
+	return BasePosOf(f.Location(), f.End())
+}
+
+// BasePosOf converts a position that is defined on a given feature to a
```
Comment (kortschak) [inline feat/feature.go]: "BasePosOf returns the position in f converted to coordinates in the deepest feature location and the reference feature. BasePosOf will panic if the feature chain is deeper than 1000 links." and implement this.

This means that position must be checked for being within f (return 0 and nil if not in f).

**Comment by kortschak** [comment]
BaseStartOf and BaseEndOf should be dropped. They save a few characters, but deny a client-cached position and location.

LGTM after these are removed and comments addressed.

**Comment by kortschak** [comment]
Please also change the commit message to match the new PR title.

**Comment by mnsmar** [comment]
Are you sure about the removal of `BaseStartOf` and `BaseEndOf`? IMO `BasePosOf(f.Location(), f.Start())` is hideous compared to `BaseStartOf(f)`. I don't get what you mean by "client-cached position and location" but it sounds to me like a very specific use case. I assume in the great majority of cases BaseStartOf and BaseEndOf are sufficient and more "user friendly".

**Comment by mnsmar** [comment]
Also isn't removing `BaseStartOf` and `BaseEndOf` similar to deleting `BaseOrientationOf` and keeping only `OrientationWithin`?

**Comment by kortschak** [comment]
> I don't get what you mean by "client-cached position and location" but it sounds to me like a very specific use case.

If I already have start from a previous call to Start() I can't use it here, and the presence of the function will drive people to it due to laziness. That's not something I want to foster.

> IMO BasePosOf(f.Location(), f.Start()) is hideous compared to BaseStartOf(f)

It should work as BasePosOf(f, f.Start()). If it doesn't it's not the right API and should be fixed.

> Also isn't removing BaseStartOf and BaseEndOf similar to deleting BaseOrientationOf and keeping only OrientationWithin?

No. OrientationWithin stops recurring at ref, while BaseOrientationOf recurs as deeply as possible. None of the position functions here provide that functionality (it would be worth having as PositionWithin).

**Comment by mnsmar** [comment]
> It should work as BasePosOf(f, f.Start()). If it doesn't it's not the right API and should be fixed."

But `f.Start()` is not defined on `f` but on `f.Location()`. This does not match what you wrote earlier

> "BasePosOf returns the position in f converted to coordinates in the deepest feature location and the reference feature. BasePosOf will panic if the feature chain is deeper than 1000 links."

The position is not in `f` but in `f.Location()`

If we want the API you suggest the description should change to the following which is a bit cumbersome IMO.

"BasePosOf returns the position defined in f.Location converted to coordinates in the deepest feature location and the reference feature. BasePosOf will panic if the feature chain is deeper than 1000 links."

**Comment by kortschak** [comment]
> BasePosOf returns the position defined in f.Location converted to coordinates in the deepest feature location and the reference feature.

I don't know what you mean by "defined in". The API I'm suggesting is really exactly what I wrote, and the way to get there is the signature I suggested unless I'm missing something. I give BasePosOf a feature and a coordinate in that feature and expect to get back a coordinate in the base feature and that base feature, or nil if the coordinate is not in the parameter feature or any feature on the path to the base. This is exactly what the comment says and what the signature describes. Currently the implementation doesn't do that, but I think that is wrong.

**Comment by mnsmar** [comment]
> ... I give BasePosOf a feature and a coordinate in that feature

The problem is that you are not giving this when you have the expression `BasePosOf(f, f.Start())` because `f.Start()` is not in `f` it's in `f.Location()`. In other words the coordinate system for `f.Start()` is `f.Location()` not `f`.

trans:   &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;100<-------------->200
gene: 0<------------------------------------------->300
In this example `trans.Start()` is 100 and is in the gene not in the transcript. The equivalent position in the transcript would be 0. So you have two possible APIs:
1. `transBaseStart := BasePosOf(gene, trans.Start()) // change coords from gene to deepest feature`
2. `transBaseStart := BasePosOf(trans, 0) // change coords from transcript to deepest feature`

you can't have what you are suggesting `transBaseStart := BasePosOf(trans, trans.Start())` (at least not with the function comment you suggested)

Also consider this `f.BasePosOf(f, f.Start()-10)`. Based on what you are saying this should not be valid because `f.Start()-10` is outside the boundaries of `f` but it is perfectly valid as long as it is within `f.Location()`

**Comment by kortschak** [comment]
> The problem is that you are not giving this when you have the expression BasePosOf(f, f.Start()) because f.Start() is not in f it's in f.Location(). In other words the coordinate system for f.Start() is f.Location() not f.

Sorry, you are correct.

```
trans:        100<-------------->200
gene: 0<------------------------------------------->300
```

> In this example trans.Start() is 100 and is in the gene not in the transcript.

True.

> You have two possible APIs:

```
    transBaseStart := BasePosOf(gene, trans.Start()) // change coords from gene to deepest feature
    transBaseStart := BasePosOf(trans, 0) // change coords from transcript to deepest feature
```

> you can't have what you are suggesting transBaseStart := BasePosOf(trans, trans.Start()) (at least not with the function comment you suggested)

Yes. I think the most useful case is the second you have above. (The base feature should be returned though.) This allows mapping from sequence slices to genome positions easily.

> Also consider this f.BasePosOf(f, f.Start()-10). Based on what you are saying this should not be valid because f.Start()-10 is outside the boundaries of f but it is perfectly valid as long as it is within f.Location()

Also true - that restriction should be dropped from what I say. Leave the bounds check for the user.

**Comment by mnsmar** [comment]
So (1) and (2) are basically the same API. Only the client uses it differently.
Given this I'm coming to my original question. Do you still think I should drop BaseStartOf and BaseEndOf given the alternatives?

Without BaseStartOf and BaseEndOf the client would have to write either
1. `start := BasePosOf(f, 0)` and `end := BasePosOf(f, f.Len())`
2. `start := BasePosOf(f.Location(), f.Start())` and `end := BasePosOf(f.Location(), f.End())`

Out of the two forms I can only see a potential benefit from caching in the second form (or maybe also for the end in the first form).

**Comment by kortschak** [comment]
In the case of hts/sam.Record Len is calculated by iterating the CIGAR string so end is O(n) where n is the number of CIGAR operations. But really, I din;t see the issue with writing what exists in 1 above (thought it does still need to return the base feature).

**Comment by mnsmar** [comment]
Sounds good. I'drop them, do the changes and update the PR

**Comment by kortschak** [comment]
Thanks for taking the time to clarify this for me.

**Comment by mnsmar** [comment]
I completed the suggested changes for `BasePositionOf()`. I also implemented a draft for the `PositionWithin()` that you mentioned in one of the comments. There are a few issues that we need to address related to edge cases. I have marked them with TODO in the code and in the tests. Let me know what you think.

**Comment by mnsmar** [comment]
Also I just noticed there might be an inconsistency in `BaseOrientationOf()`. The description says "returns .... the reference feature, which may be the feature itself if it is not an Orienter **or has a nil Location**. Check the last test case in `orientationTests`. Although `freeOri` has a `nil` location the function returns `nil` instead of `freeOri`. I think the test describes the expected behavior better. The caveat with this though is that if you reach `nil` in two independent cases you cannot know if the underlying trees match and whether the returned orientations are comparable.
e.g.

``` go
ori1, ref1 := BaseOrientationOf(f1)
ori2, ref2 := BaseOrientationOf(f2)
if ref1 == ref2 {
     // in here you don't know if ori1 is comparable to ori2
     // you would have to check ref1, ref2 for nil and treat them accordingly
}
```

**Comment by kortschak** [comment]
Ooops, wandered off into commit land. github codereview is horrible.

**Comment by kortschak** [comment]
> Also I just noticed there might be an inconsistency in BaseOrientationOf(). The description says "returns .... the reference feature, which may be the feature itself if it is not an Orienter or has a nil Location. Check the last test case in orientationTests. Although freeOri has a nil location the function returns nil instead of freeOri. I think the test describes the expected behavior better.

What do you want to do about it? If possible it would be nice to have symmetry between the position and orientation functions.

> The caveat with this though is that if you reach nil in two independent cases you cannot know if the underlying trees match and whether the returned orientations are comparable.

I think this is OK. feat.Feature(nil) is not equal to feat.Feature(nil). Maybe this should be more clearly documented or an example provided.

Both of these things would be in a separate PR.

**Comment by mnsmar** [comment]
> What do you want to do about it? If possible it would be nice to have symmetry between the position and orientation functions.

I was thinking the same thing about the symmetry and that's why I mentioned it here. I think that the functions although they look alike have a very subtle difference. Namely `BaseOrientationOf` starts by definition one step deeper than `BasePositionOf`. If we want a completely symmetric implementation to `BasePositionOf` then we should define `BaseOrientationOf` as `BaseOrientationOf(f Feature, ori Orientation)`. `ori` would be an orientation defined on `f` and not the orientation of `f`.

And there is another thing to note. With the current implementation of `BasePositionOf` if we change the start of `freeOri` to 10 it does not make any difference. The reason is that `BasePositionOf` stops early due to `freeOri.Location() == nil`. As it is now, the base position of the start of `freeOri` is 0 and not 10 (`BasePositionOf(freeOri, 0)`). If we want to allow `BasePositionOf` to reach a `nil` then it will _always_ reach a `nil`. Note that the only thing preventing `BaseOrientationOf` from always reaching a nil is the `Orienter` check.

``` go
freeOri = ori{
    nonOri: nonOri{
        start: 0, end: 100,
        name: "frag",
        desc: "fragment",
    },
    orient: feat.Reverse,
}
```

To sum up, my feeling is that we should leave the behavior as is and maybe change the function naming a bit to avoid confusions.

**Comment by kortschak** [comment]
I'll have a think about this.

**Comment by kortschak** [comment]
> Note that the only thing preventing BaseOrientationOf from always reaching a nil is the Orienter check.

The analogue for positions would be that he position is outside the location feature. I think this is a meaningful test.

**Comment by mnsmar** [comment]
It could be a meaningful test but all legit trees (locations within bounds) would still reach the nil.

**Comment by kortschak** [comment]
Yeah, you're right. The full analogue then is that if the location's location is nil, the the location should be returned, i.e. stop before the nil, and include the above test.

**Comment by mnsmar** [comment]
Please check and let me know if anything is missing.

File: feat/feature.go [hunk 4]
```diff
@@ -105,6 +105,45 @@ type Collection interface {
 	Location() Feature
 }
 
+// BasePositionOf returns the position in f converted to coordinates in the
+// deepest non nil feature location and the reference feature which may be the
+// feature itself if it has a nil Location.
+// BasePositionOf will panic if the feature chain is deeper than 1000 links.
+func BasePositionOf(f Feature, position int) (int, Feature) {
+	if f == nil {
```
Comment (mnsmar) [inline feat/feature.go]: Should I also let this panic? I think that the return kind of makes sense here.

Comment (kortschak) [inline feat/feature.go]: Yes, it should panic.

**Comment by kortschak** [comment]
LGTM with minor comment.

I'd like to come back to the issue raised above, but I don't have the time right now. I'll file an issue to resolve the subtleties of these functions (or decide they are resolved).

**Comment by mnsmar** [comment]
Related to #30 
It seems that the true analog between the two conditions is that at some point in the hierarchy you reach an "invalid" location. That would be a `nil` for `BasePositionOf` and a `nil` or a `NonOrienter` for `BaseOrientationOf`. If you think it this way then logically an "invalid" location cannot serve as a reference location to be returned. So the functions should return the last valid location.

Bottom-line we should let both functions reach the invalid location consuming the appropriate properties (starts and orientations) as they move down the hierarchy and they should return the last _valid_ location after they have consumed it. The _valid_ locations can be used by the client to compare between independent calls (_valid_ locations should be equal for the returned values to be comparable).

**Comment by kortschak** [comment]
SGTM

**Comment by mnsmar** [comment]
OK. So I'll only implement BasePositionOf() in this PR and leave the BaseOrientationOf for another PR that will also close the issue.

**Comment by kortschak** [comment]
Yep, that's the plan.

**Comment by kortschak** [comment]
Are you ready for me to merge this?

**Comment by mnsmar** [comment]
Not yet. Let me make the changes above and I'll let you know

File: feat/feature.go [hunk 5]
```diff
@@ -105,6 +105,42 @@ type Collection interface {
...
+
+// PositionWithin returns the position in f converted to coordinates in a
+// given reference feature and a boolean value indicating if the conversion
+// was performed successfully. If a false value is returned then no assumption
+// can be made about the position.
+// PositionWithin will panic if the feature chain is deeper than 1000 links.
+func PositionWithin(f, ref Feature, position int) (int, bool) {
+	for n := 0; n < 1000; n++ {
+		if f == ref {
+			return position, f != nil
```
Comment (mnsmar) [inline feat/feature.go:131]: Review: If both of them are nil it returns false.

File: feat/feature_test.go [hunk 6]
```diff
@@ -288,3 +288,89 @@ func (s *S) TestOrientationWithin(c *check.C) {
...
+		pos:       20,
+		basePos:   75,
+		posWithin: 75,
+		ref:       chrom1,
+	},
+	{
+		f:         freeOri,
+		pos:       0,
+		basePos:   10,
+		posWithin: 0,
```
Comment (mnsmar) [inline feat/feature_test.go:336]: Review: Had to add posWithin cause now `PositionWIthin()` has a slightly different meaning that `BasePositionOf()`

File: feat/feature_test.go [hunk 7]
```diff
@@ -288,3 +288,89 @@ func (s *S) TestOrientationWithin(c *check.C) {
...
+		pos, ok := feat.PositionWithin(t.f, t.ref, t.pos)
+		c.Check(pos, check.Equals, t.posWithin)
+		c.Check(ok, check.Equals, true)
+	}
+
+	// Check unorthodox tree structures.
+	_, ok := feat.PositionWithin(opA, chrom2, 10)
+	c.Check(ok, check.Equals, false)
+	_, ok = feat.PositionWithin(nil, nil, 10)
+	c.Check(ok, check.Equals, false)
```
Comment (mnsmar) [inline feat/feature_test.go:370]: Review: nil vs nil returns false

**Comment by mnsmar** [comment]
I'm done with the changes but because I'm tired please be extra careful with the review. I've added some comments on interesting points. If you think all is good feel free to merge.

File: feat/feature.go (hunk 5 above)
Comment (kortschak) [inline feat/feature.go:131]: Yes, I think this is OK.

File: feat/feature.go [hunk 8]
```diff
@@ -105,6 +105,42 @@ type Collection interface {
...
+	}
+	panic("feat: feature chain too long")
+}
+
+// PositionWithin returns the position in f converted to coordinates in a
+// given reference feature and a boolean value indicating if the conversion
+// was performed successfully. If a false value is returned then no assumption
+// can be made about the position.
+// PositionWithin will panic if the feature chain is deeper than 1000 links.
+func PositionWithin(f, ref Feature, position int) (int, bool) {
```
Comment (kortschak) [inline feat/feature.go]: Please add (pos int, ok bool) for the returns; this indicates what the boolean means.

File: feat/feature_test.go (hunk 7 above)
Comment (kortschak) [inline feat/feature_test.go:370]: Yes.

**Comment by mnsmar** [comment]
done

File: feat/feature.go [hunk 9]
```diff
@@ -105,6 +105,42 @@ type Collection interface {
...
+			f = f.Location()
+			continue
+		}
+		return position, f
+	}
+	panic("feat: feature chain too long")
+}
+
+// PositionWithin returns the position in f converted to coordinates in a
+// given reference feature and a boolean value indicating if the conversion
```
Comment (kortschak) [inline feat/feature.go]: "... and a boolean indicating whether f can be located relative to ref." and drop the following sentence.

File: feat/feature_test.go (hunk 6 above)
Comment (kortschak) [inline feat/feature_test.go:336]: OK

**Comment by kortschak** [comment]
LGTM after doc change. No hurry.

**Comment by mnsmar** [comment]
Fixed the doc

**Comment by kortschak** [comment]
Thanks.
